lib/services/local/inventory.go

```diff
@@ -146,4 +146,38 @@ func (s *PresenceService) CompareAndSwapInstance(ctx context.Context, instance t
 	return value, nil
 }
 
+// UpsertInstance creates or updates an instance resource.
+func (s *PresenceService) UpsertInstance(ctx context.Context, instance types.Instance) error {
+	if err := instance.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// instance resource expiry is calculated relative to LastSeen and/or the longest living
+	// control log entry (whichever is further in the future).
+	if instance.GetLastSeen().IsZero() || instance.Expiry().IsZero() {
+		instance.SetLastSeen(s.Clock().Now().UTC())
+		instance.SyncLogAndResourceExpiry(apidefaults.ServerAnnounceTTL)
+	}
```

Comment:
Does this belong here in PresenceService, in the CheckAndSetDefaults method, or in (*Server).UpsertInstance?

Reply:
Tough to say.  The thing is that I want this to happen after `CheckAndSetDefaults`, but *only* on backend insertion, not in other places that `CheckAndSetDefaults` might get called.

Honestly, I think we'd be better off if we separated "insertion" checks from "usage" checks more generally... I don't particularly like the idea of modifying meaningful fields on read outside of an explicit migration.

In some of the old upgrade system work I did that never landed I used `OnCreateChecks` for a number of resources and included additional checks that I anticipated would cause problems during an auth downgrade (such as asserting that the value of an enum field was recognized).  I liked that pattern, tho in retrospect I think it wasn't general enough... something like `OnWrite() error` might have been better... something that says "this is like `CheckAndSetDefaults` but we only do it once when adding this to the backend"... idk, naming is hard.  Regardless, I think something like this will be necessary if we want to move toward more heavy use of generics with less per-resource boilerplate.

Which I guess is a long-winded way of saying that I _think_ this is the right place for now, but am not 100% happy with it.